packages/app/src/sandbox/worker/index.ts

```diff
@@ -0,0 +1,111 @@
+import { invariant } from 'outvariant';
+
+import { debug, getServiceWorker, preventStaleTermination } from './utils';
+import {
+  CHANNEL_NAME,
+  IPreviewInitMessage,
+  IPreviewReadyMessage,
+  IPreviewResponseMessage,
+  IWorkerInitMessage,
+} from './types';
+import { DeferredPromise } from './promise';
+
+// Create a message channel for communication with the Service Worker.
+const workerChannel = new MessageChannel();
+
+const workerReadyPromise = new DeferredPromise<ServiceWorker>();
+
+workerReadyPromise.then(worker => {
+  debug('[relay] worker is ready, initializing MessageChannel...');
+
+  // Always post the initial MessageChannel message to the worker
+  // as soon as the worker is ready. This is done once.
+  const workerInitMessage: IWorkerInitMessage = {
+    $channel: CHANNEL_NAME,
+    $type: 'worker/init',
+  };
+  worker.postMessage(workerInitMessage, [workerChannel.port2]);
+
+  return worker;
+});
+
+const parentPortPromise = new DeferredPromise<MessagePort>();
+window.addEventListener(
```

Comment:
This could potentially be a breaking change, but not sure if it would have a real impact?